subt/main.py

```diff
@@ -469,6 +469,32 @@ def on_pose2d(self, timestamp, data):
         self.xyz = x, y, z
         self.trace.update_trace(self.xyz)
 
+    def on_pose3d(self, timestamp, data):
+        xyz, rot = data
+        ypr = quaternion.euler_zyx(rot)
+        pose = (xyz[0], xyz[1], ypr[0])
+        if self.last_position is not None:
+            self.is_moving = (self.last_position != pose)
+            dist = math.hypot(pose[0] - self.last_position[0], pose[1] - self.last_position[1])
+            direction = ((pose[0] - self.last_position[0]) * math.cos(self.last_position[2]) +
+                         (pose[1] - self.last_position[1]) * math.sin(self.last_position[2]))
+            if direction < 0:
+                dist = -dist
+        else:
+            dist = 0.0
+        self.last_position = pose
+        self.traveled_dist += dist
+        x, y, z = xyz
+        self.last_send_time = self.bus.publish('pose2d', [round(x * 1000), round(y * 1000),
+                                    round(math.degrees(self.yaw) * 100)])
+        if self.virtual_bumper is not None:
+            if self.is_virtual:
+                self.virtual_bumper.update_pose(timedelta(seconds=self.sim_time_sec), pose)
+            else:
+                self.virtual_bumper.update_pose(self.time, pose)
+        self.xyz = tuple(xyz)
```

Comment:
Some pieces of the code, such as when reporting artifacts, add `self.offset` to `self.xyz`. To make that fully compatible, the easiest thing is probably to subtract `self.offset` somewhere here.

Reply:
Hmm ... sigh ... I am not sure if I want to patch it again or drop `self.offset` completely ... and it will surely break things ...

Reply:
Honestly, I would a) make sure we have `self.offset` before driving, b) subtract it from `xyz` in the beginning of this function, c) think about a follow-up independent cleanup pass.

For example, this new `on_pose3d` allows using estimates from the localization module. Which in turn means that once we have updated configs for all robots (sigh), `on_pose2d` becomes obsolete and can be deleted. That, in turn, may make removing `self.offset` easier.

Reply:
OK, added attempt to handle offset for `on_pose3d` (not tested yet). `self.offset` is now set to undefined - for real robots it has to be passed from outside or from configuration (I would not support it any more - for two month at least) and for virtual it is updated on the very beginning. The request for origin can fail after returning from the tunnel, but then the `self.offset` is already set.

Reply:
p.s. surely the `pose3d` can come sooner before the origin is defined ... when replaying the only log from AWS I see:
```
  File "M:\git\osgar\subt\main.py", line 476, in on_pose3d
    assert False, xyz  # hack
AssertionError: [-6.0, 5.0, 0.125]
```
(added debug assert before return to ignore incoming `pose3d`)

Reply:
Well, the localization module does not publish `pose3d` until it has received the origin so the value is valid regardless of origin being also received within the subt main application.
https://github.com/robotika/osgar/blob/70fc835d7c146532207d5e1b0fcc66dd4a17a28c/subt/visual_odometry.py#L951-L956
https://github.com/robotika/osgar/blob/70fc835d7c146532207d5e1b0fcc66dd4a17a28c/subt/odoimuloc.py#L60-L65
My proposal would be to not deal with origin in the main application and instead leave it for the localization modules.

Reply:
I can confirm that the received `pose3d` was already corrected for origin. Nevertheless I agree with @jisa to close this "now" and refactor pose3d in another PR then "introduction of X4", OK?

Reply:
Can be... the question is - what needs to be done to merge this PR? I see @jisa is requesting changes.